File: arkcompiler/ets_runtime/ecmascript/platform/unix/map.cpp
/*
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ecmascript/platform/map.h"

#include <cerrno>
#include <sys/mman.h>
#include <unistd.h>

#include "ecmascript/log_wrapper.h"
#include "ecmascript/mem/mem.h"
#include "ecmascript/platform/os.h"

namespace panda::ecmascript {
MemMap PageMap(size_t size, int prot, size_t alignment)
{
    ASSERT(size == AlignUp(size, PageSize()));
    ASSERT(alignment == AlignUp(alignment, PageSize()));
    size_t allocSize = size + alignment;
    void *result = mmap(nullptr, allocSize, prot, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (reinterpret_cast<intptr_t>(result) == -1) {
        LOG_ECMA(FATAL) << "mmap failed with error code:" << errno;
    }
    if (alignment != 0) {
        auto alignResult = AlignUp(reinterpret_cast<uintptr_t>(result), alignment);
        size_t leftSize = alignResult - reinterpret_cast<uintptr_t>(result);
        size_t rightSize = alignment - leftSize;
        void *alignEndResult = reinterpret_cast<void *>(alignResult + size);
        munmap(result, leftSize);
        munmap(alignEndResult, rightSize);
        result = reinterpret_cast<void *>(alignResult);
    }
    return MemMap(result, size);
}

void PageUnmap(MemMap it)
{
    munmap(it.GetMem(), it.GetSize());
}

MemMap MachineCodePageMap(size_t size, int prot, size_t alignment)
{
    MemMap memMap = PageMap(size, prot, alignment);
    PageTag(memMap.GetMem(), memMap.GetSize(), PageTagType::MACHINE_CODE);
    return memMap;
}

void MachineCodePageUnmap(MemMap it)
{
    PageClearTag(it.GetMem(), it.GetSize());
    PageProtect(it.GetMem(), it.GetSize(), PAGE_PROT_NONE);
    PageUnmap(it);
}

void PageRelease(void *mem, size_t size)
{
    madvise(mem, size, MADV_DONTNEED);
}

void PagePreRead(void *mem, size_t size)
{
    madvise(mem, size, MADV_WILLNEED);
}

void PageTag(void *mem, size_t size, PageTagType type)
{
    const char *tag = GetPageTagString(type);
    PrctlSetVMA(mem, size, tag);
}

void PageClearTag(void *mem, size_t size)
{
    PrctlSetVMA(mem, size, nullptr);
}

void PageProtect(void *mem, size_t size, int prot)
{
    int ret = mprotect(mem, size, prot);
    if (ret != 0) {
        LOG_ECMA(ERROR) << "PageProtect change to " << prot << " failed, error code is " << errno;
    }
}

size_t PageSize()
{
    return getpagesize();
}
}  // namespace panda::ecmascript

File: arkcompiler/ets_runtime/test/perform/map/map.ts
# Copyright (c) 2023 Huawei Device Co., Ltd.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

declare function print(arg:any) : string;
declare interface ArkTools {
    timeInUs(arg:any):number
}

function testSet() {
    let myMap = new Map();
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 100_000; i++) {
        myMap.set(i, i);
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Set:\t" + String(time) + "\tms");
}

function testClear() {
    let myMap = new Map();
    for (let i = 0; i < 100_000; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 100_000; i++) {
        myMap.clear();
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Clear:\t" + String(time) + "\tms");
}

function testDelete() {
    let myMap = new Map();
    for (let i = 0; i < 100_000; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.delete(i);
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Delete:\t" + String(time) + "\tms");
}

function testHas() {
    let myMap = new Map();
    for (let i = 0; i < 100_000; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.has(i);
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Has:\t" + String(time) + "\tms");
}

function setElements(value, key, map) {}
function testForEach() {
    let myMap = new Map();
    for (let i = 0; i < 10; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.forEach(setElements);
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map ForEach:\t" + String(time) + "\tms");
}

function testKeys() {
    let myMap = new Map();
    for (let i = 0; i < 10; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.keys();
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Keys:\t" + String(time) + "\tms");
}

function testValues() {
    let myMap = new Map();
    for (let i = 0; i < 10; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.values();
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Values:\t" + String(time) + "\tms");
}

function testEntries() {
    let myMap = new Map();
    for (let i = 0; i < 10; i++) {
        myMap.set(i, i);
    }
    let start = ArkTools.timeInUs();
    for (let i = 0; i < 1_000_000; i++) {
        myMap.entries();
    }
    let end = ArkTools.timeInUs();
    let time = (end - start) / 1000
    print("Map Entries:\t" + String(time) + "\tms");
}

testSet();
testClear();
testDelete();
testHas();
testForEach();
testKeys();
testValues();
testEntries();
